Core/src/IfcGeometryConverter/ProfileConverter.h

```diff
@@ -517,499 +496,361 @@ namespace OpenInfraPlatform {
-								}
-								paths.push_back(outer_loop);
-								paths.push_back(inner_loop);
-							}
-							return;
-						}
-
-						// IfcRoundedRectangleProfileDef SUBTYPE OF IfcRectangleProfile 
-						std::shared_ptr<typename IfcEntityTypesT::IfcRoundedRectangleProfileDef> rounded_rectangle =
-							std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcRoundedRectangleProfileDef>(rectangle_profile);
-						if(rounded_rectangle) {
-							if(rounded_rectangle->RoundingRadius) {
-								double rr = rounded_rectangle->RoundingRadius * length_factor;
-								addArc(outer_loop, rr, 0, M_PI_2, x * 0.5 - rr, y * 0.5 - rr);
-								addArc(outer_loop, rr, M_PI_2, M_PI_2, -x * 0.5 + rr, y * 0.5 - rr);
-								addArc(outer_loop, rr, M_PI, M_PI_2, -x * 0.5 + rr, -y * 0.5 + rr);
-								addArc(outer_loop, rr, 3 * M_PI_2, M_PI_2, x * 0.5 - rr, -y * 0.5 + rr);
-								paths.push_back(outer_loop);
-							}
-							return;
-						}
-
-						// Else it's a standard rectangle
-						outer_loop.push_back(carve::geom::VECTOR(-x * 0.5, -y * 0.5));
-						outer_loop.push_back(carve::geom::VECTOR(x * 0.5, -y * 0.5));
-						outer_loop.push_back(carve::geom::VECTOR(x * 0.5, y * 0.5));
-						outer_loop.push_back(carve::geom::VECTOR(-x * 0.5, y * 0.5));
-						paths.push_back(outer_loop);
-						return;
-					}
+				std::shared_ptr<typename IfcEntityTypesT::IfcRectangleProfileDef> rectangle_profile = std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcRectangleProfileDef>(profileDef);
+				if (rectangle_profile) {
+					convertIfcRectangleProfileDef(EXPRESSReference<typename IfcEntityTypesT::IfcRectangleProfileDef>(rectangle_profile), paths, outer_loop);
 				}
 
 				// (2/10) IfcTrapeziumProfileDef SUBTYPE OF IfcParametrizedProfileDef
-				std::shared_ptr<typename IfcEntityTypesT::IfcTrapeziumProfileDef> trapezium = std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcTrapeziumProfileDef>(profileDef);
-				if(trapezium) {
-					if(trapezium->BottomXDim && trapezium->TopXDim && trapezium->TopXOffset && trapezium->YDim) {
-						double xBottom = trapezium->BottomXDim * length_factor;
-						double xTop = trapezium->TopXDim * length_factor;
-						double xOffset = trapezium->TopXOffset * length_factor;
-						double y = trapezium->YDim * length_factor;
-						outer_loop.push_back(carve::geom::VECTOR(-xBottom * 0.5, -y * 0.5));
-						outer_loop.push_back(carve::geom::VECTOR(xBottom * 0.5, -y * 0.5));
-						outer_loop.push_back(carve::geom::VECTOR(-xBottom * 0.5 + xOffset + xTop, y * 0.5));
-						outer_loop.push_back(carve::geom::VECTOR(-xBottom * 0.5 + xOffset, y * 0.5));
-						paths.push_back(outer_loop);
-					}
-					return;
+				EXPRESSReference<typename IfcEntityTypesT::IfcTrapeziumProfileDef> trapezium = std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcTrapeziumProfileDef>(profileDef);
+				if (trapezium) {
+					convertIfcTrapeziumProfileDef(EXPRESSReference<typename IfcEntityTypesT::IfcTrapeziumProfileDef>(trapezium), paths, outer_loop);

```

Comment:
```suggestion
				if (profileDef.isOfType<typename IfcEntityTypesT::IfcTrapeziumProfileDef>()) {
					convertIfcTrapeziumProfileDef(profileDef.as<typename IfcEntityTypesT::IfcTrapeziumProfileDef>(), paths, outer_loop);
```
And similarly everywhere else.

Reply:
Resolved: afab19f and e76c77e